### PR TITLE
Apply join rate limiter outside the lineariser

### PR DESCRIPTION
**Pull request by DMRobertson**
I came across this logic; got confused, then tried to tidy it up. I found this much easier to reason about---doing the rate limit check up front.

At some point, someone should draw out a flowchart of the various membership-updating functions here. It's very confusing.

Commitwise reviewable.

### CODE REVIEW

File: synapse/handlers/room_member.py [hunk 1]
```diff
@@ -618,6 +604,25 @@ async def update_membership(
...
+                    requester.user.to_string(),
+                    room_id,
+                )
+                if current_membership != Membership.JOIN:
+                    await self._join_rate_limiter_local.ratelimit(requester)
+                    await self._join_rate_per_room_limiter.ratelimit(
+                        requester, key=room_id, update=False
+                    )
+            elif action == Membership.INVITE:
+                await self.ratelimit_invite(requester, room_id, target.to_string())
```
Comment (clokep) [inline synapse/handlers/room_member.py:624]: Nice consolidation here. 👍 

I wondered why we couldn't remove `ratelimit` from other spots, but we pass it into `handle_new_client_event`.

**Review by clokep — APPROVED** [review]
Good w/ me if tests look good.

File: synapse/handlers/room_member.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/handlers/room_member.py:624]: I had the same thought (sorry, should have left a comment to this effect), and decided that I was too scared to see if the ratelimiting logic was further disentangleable.

Comment (clokep) [reply synapse/handlers/room_member.py:624]: Completely fair, I think this is a solid improvement.

**Pull request merged**